fix some major memory leak (+ some small leak...)

qyx268 opened the pull request:
fix #161 

Before merging into the master, can @jaehongpark00 pull the changes from this `free_mem` branch and test with the current photonCon run on `nefertem` to see if the RAM still rises?

qyx268 commented:
hmm... currently, some of the test failed at test_integration_features when using FFTW. This is probably related to #179. I'm wondering why this didn't show up before...

I'll come back to this later.

qyx268 commented:
er... the problem was solved after fixing a bug which i dont think it's responsble... anyhow. can someone please do a review?  @steven-murray @BradGreig

BradGreig commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -1688,12 +1688,30 @@ LOG_SUPER_DEBUG("freed fftw boxes");
 
+    if (prev_redshift < 1){
+        free(previous_ionize_box->z_re_box);
+        if (flag_options->USE_MASS_DEPENDENT_ZETA) {
```
Maybe (flag_options->USE_MASS_DEPENDENT_ZETA && flag_options->USE_MINI_HALOS)

BradGreig commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -1688,12 +1688,30 @@ LOG_SUPER_DEBUG("freed fftw boxes");
         }
     }
 
+    if(INIT_ERFC_INTERPOLATION) {
```
No, this shouldn't be done like this (well, at least not currently).

These arrays are generated once per session, thus they'll need to be freed differently. At the top of the file (line 194), where these are first allocated, instead do:
if(!flag_options->USE_MASS_DEPENDENT_ZETA && INIT_ERFC_INTERPOLATION) {

These are only used for the old, T_VIR dependence

qyx268 replied:
Can I confirm that you are saying to add `!flag_options->USE_MASS_DEPENDENT_ZETA` to both this line and 194 rather than saying freeing these two arrays here is wrong? I agree this is only used when USE_MASS_DEPENDENT_ZETA==False

BradGreig replied:
Ah, yeah, sorry for not being clear. I was saying to add `!flag_options->USE_MASS_DEPENDENT_ZETA` to line 194 (as you have done). However, you cannot free `ERFC_VALS` or `ERFC_VALS_DIFF`. Thus, I would remove the freeing of these. These are populated once per session, thus freeing them will cause segmentation faults.

We'll need to come up with a better way to free these. However, adding `!flag_options->USE_MASS_DEPENDENT_ZETA` to line 194 will at least ensure this isn't populated in most cases.

qyx268 replied:
Ah, I see what you mean. I didn't realize `INIT_ERFC_INTERPOLATION` is a global parameter rather than an input parameter. I will remove the freeing then

BradGreig commented on src/py21cmfast/wrapper.py:
```diff
@@ -2366,6 +2366,9 @@ def run_coeval(
             photon_nonconservation_data = None
 
+        if flag_options.USE_TS_FLUCT and user_params.USE_INTERPOLATION_TABLES:
+            lib.FreeTsInterpolationTables(flag_options())
```
This is only for the `run_coeval` instance. You need to ensure this is also done for the `run_lightcone` as well.

qyx268 commented:
The recent commits should also fix #182

qyx268 commented:
for the record, sometime this line in `SpinTemperatureBox.c`
`(this_spin_temp->first_box || (fabs(initialised_redshift - perturbed_field_redshift) > 0.0001` is True at multiple redshifts which results in executing `initialiseSigmaMInterpTable`, `initialise_Nion_Ts_spline` and `initialise_SFRD_spline` and allocating memory for interpolation tables multiple times. Therefore, I added a check to see if the pointer is NULL before allocating. 

However, this is not easy for local static pointers. Therefore, I moved init_heat() out of the aforementioned check and do destruct_heat() for each SpinTemperatureBox call. This is not a heavy overhead since the tables are very small.

qyx268 commented:
@jaehongpark00's MCMC run shows much better improvement for RAM. There are only some small leakages due to repeated interpolation table allocation, which is now fixed.

@BradGreig  could you please see the change you requested and let me know if the new implementations make sense to you.
